pipeline/matching/rankers/variety.py

```diff
@@ -1,17 +1,31 @@
-from typing import Iterator
+from typing import Iterator, List
 
 from pipeline.matching.core import MatchRanker
 from pipeline.types import Match, MatchingInput
 
 RANKER_VARIETY = "varietyRanker"
 
 
+def dup_gencheck(recentMatches: List[Match], match: Match) -> bool:
+    for recent in recentMatches:  # iterate list of Match's
+        if (
+            match.metadata.generator == recent.metadata.generator
+        ):  # check if the same generators are used
+            if match.users.intersection(
+                recent.users
+            ):  # check if any of the users in a match pair reappears in a recent match
+                return True  # YES, there is a duplicate
+    return False  # NO duplicates
+
+
 class VarietyRanker(MatchRanker):
     def __init__(self):
         super().__init__(name=RANKER_VARIETY)
 
     def rank(
         self, inp: MatchingInput, proposed: Iterator[Match]
     ) -> Iterator[Match]:
-        # TODO: Implement your ranker
-        yield from proposed
+        inp.logger.info("David was here!")
+        yield from sorted(
+            proposed, key=lambda match: dup_gencheck(inp.recent_matches, match)
+        )
```

Comment:
Generators for matches behaved a little differently than I expected and tested for, so every time we look at a generator the value gets taken out of the generator.

However, we are only looking at one specific match at a time and the MatchingInput, which is not a generator so we are good! 

Reference: **https://github.com/scarletstudio/butterfly/pull/248/files#r917440831**